### ISSUE TITLE
Feature : attrs should accept a callback (not just an object)

### ISSUE THREAD

**Issue by sylvain-hamel** [styled-components/styled-components#1725]
**Exemple:**

```
const addNewProps = props => // return an object with an arbitrary number of props;
const MyComponent = styled(Something).attrs(props=> addNewProps(props));

```
**What will it allow you to do that you can't do today?**
Use `attrs` to introduce a variable number of props, not a list of _known_ props.

**How will it make current work-arounds straightforward?**
Avoid unnecessary wrappers that just pass on some props to the rendered component.

**Comment by clayne11** [comment]
It will also allow us to add strong-typing for `attrs` in Typescript which isn't possible with the current API.

**Comment by probablyup** [comment]
Closed via #2200
